Mathlib/Topology/MetricSpace/BundledFun.lean

```diff
@@ -0,0 +1,181 @@
+  /-- The underlying binary function mapping into a linearly ordered additive monoid. -/
+  toFun : X → X → R
+  /-- A pseudometric must take identical elements to 0. -/
+  refl' x : toFun x x = 0
+  /-- A pseudometric must be symmetric. -/
+  symm' x y : toFun x y = toFun y x
+  /-- A pseudometric must respect the triangle inequality. -/
+  triangle' x y z : toFun x z ≤ toFun x y + toFun y z
+
+namespace PseudoMetric
+
+instance : FunLike (PseudoMetric X R) X (X → R) where
+  coe := PseudoMetric.toFun
+  coe_injective' _ := by aesop
+
+variable (d : PseudoMetric X R)
+
+@[simp, norm_cast]
+lemma coe_mk (d : X → X → R) (refl symm triangle) : mk d refl symm triangle = d := rfl
+
+lemma mk_apply (d : X → X → R) (refl symm triangle) (x y : X) :
+    mk d refl symm triangle x y = d x y :=
+  rfl
+
+@[simp]
+protected lemma refl (x : X) : d x x = 0 := d.refl' x
+protected lemma symm (x y : X) : d x y = d y x := d.symm' x y
+protected lemma triangle (x y z : X) : d x z ≤ d x y + d y z := d.triangle' x y z
+
+instance : LE (PseudoMetric X R) := ⟨fun d d' ↦ ⇑d ≤ d'⟩
+
+@[simp, norm_cast]
+protected lemma coe_le_coe_iff_le {d d' : PseudoMetric X R} :
+    (d : X → X → R) ≤ d' ↔ d ≤ d'  :=
+  Iff.rfl
+
+instance : PartialOrder (PseudoMetric X R) where
+  le_refl := by simp [← PseudoMetric.coe_le_coe_iff_le]
+  le_trans f _ _ := le_trans (a := ⇑f)
+  le_antisymm f g := by simpa [← PseudoMetric.coe_le_coe_iff_le, PseudoMetric.ext_iff]
+    using le_antisymm
+
+instance : Bot (PseudoMetric X R) where
+  bot.toFun := 0
+  bot.refl' _ := rfl
+  bot.symm' _ _ := rfl
+  bot.triangle' _ _ _ := by simp
+
+@[simp, norm_cast]
+lemma coe_bot : ⇑(⊥ : PseudoMetric X R) = 0 := rfl
+
+@[simp]
+protected lemma bot_apply (x y : X) :
+    (⊥ : PseudoMetric X R) x y = 0 :=
+  rfl
+
+instance : Max (PseudoMetric X R) where
+  max d d' := {
+    toFun := fun x y ↦ (d x y) ⊔ (d' x y)
+    refl' _ := by simp
+    symm' x y := by simp [d.symm, d'.symm]
+    triangle' := by
+      intro x y z
+      simp only [sup_le_iff]
+      refine ⟨(d.triangle x y z).trans ?_, (d'.triangle x y z).trans ?_⟩ <;>
+      apply add_le_add <;> simp
+  }
+
+@[simp]
+protected lemma sup_apply (d' : PseudoMetric X R) (x y : X) :
+    (d ⊔ d') x y = (d x y) ⊔ (d' x y) := rfl
+
+instance : SemilatticeSup (PseudoMetric X R) where
+  sup := max
+  le_sup_left := by simp [← PseudoMetric.coe_le_coe_iff_le, Pi.le_def]
+  le_sup_right := by simp [← PseudoMetric.coe_le_coe_iff_le, Pi.le_def]
+  sup_le _ _ _ := fun h h' _ _ ↦ sup_le (h _ _) (h' _ _)
+
+section OrderBot
+
+variable [AddLeftStrictMono R]
+
+protected lemma nonneg (x y : X) : 0 ≤ d x y := by
+  by_contra! H
+  have : d x x < 0 := by
+    calc d x x ≤ d x y + d y x := d.triangle' x y x
+      _ < 0 + 0 := by refine add_lt_add H (d.symm x y ▸ H)
+      _ = 0 := by simp
+  exact this.not_le (d.refl' x).ge
+
+instance : OrderBot (PseudoMetric X R) where
+  bot_le f _ _ := f.nonneg _ _
+
+lemma finsetSup_apply {Y : Type*} {f : Y → PseudoMetric X R}
```

Comment:
Can we have `@[simp, norm_cast] lemma coe_finsetSup`?

Reply:
I'll add it, but we don't have any good lemmas for dealing with Finset.sup' application over funlikes.